clippy_lints/src/methods/mod.rs

```diff
@@ -951,6 +951,38 @@ declare_clippy_lint! {
     "suspicious usage of map"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for `MaybeUninit::uninit().assume_init()`.
+    ///
+    /// **Why is this bad?** For most types, this is undefined behavior.
+    ///
+    /// **Known problems:** For now, we accept empty tuples and tuples / arrays
+    /// of `MaybeUninit`. There may be other types that allow uninitialized
+    /// data, but those are not yet rigorously defined.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // Beware the UB
+    /// use std::mem::MaybeUninit;
+    ///
+    /// let _: usize = unsafe { MaybeUninit::uninit().assume_init() };
```

Comment:
Is there a correct way to do the same? Or ist this just a "Don't do this at all" lint?

Reply:
The correct way is hard to know in the general case. Perhaps for `Default` types we could suggest using `Default::default` instead. But that might be misleading if the user made a type error and wanted to initialize a uninit-compatible type instead.